Use `Wunused-crate-dependencies` for the compiler

nnethercote opened the pull request:
An implementation of https://github.com/rust-lang/compiler-team/issues/844.

r? @jieyouxu

nnethercote committed:
Use `-Wunused_crate_dependencies` for compiler crates.

It's very useful. There are some false positives involving integration
tests in `rustc_pattern_analysis` and `rustc_serialize`. There is also a
false positive involving `rustc_driver_impl`'s
`rustc_randomized_layouts` feature. And I removed a `rustc_span` mention
in a doc comment in `rustc_log` because it wasn't integral to the
comment but caused a dev-dependency.

nnethercote committed:
Convert `rustc_serialize` integration tests to unit tests.

Because (a) the vast majority of compiler tests are unit tests, and (b)
this works better with `unused_crate_dependencies`.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13629920858/job/38095262582) [(plain)](https://github.com/rust-lang/rust/commit/9c4840fa061f8d14fa1ef2bafbc0d95f4620d063/checks/38095262582/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04
ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
  make \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 2.880 Building wheels for collected packages: reuse
#12 2.881   Building wheel for reuse (pyproject.toml): started
#12 3.109   Building wheel for reuse (pyproject.toml): finished with status 'done'
#12 3.110   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=0c2fd2aaf7b0bf8d6e131220aff14712a774c2ca462f3204d25460cbcf610b63
#12 3.110   Stored in directory: /tmp/pip-ephem-wheel-cache-zj2e_q97/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#12 3.113 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#12 3.511 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#12 3.512 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 4.052 Collecting virtualenv
#12 4.052 Collecting virtualenv
#12 4.106   Downloading virtualenv-20.29.2-py3-none-any.whl (4.3 MB)
#12 4.259      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 28.4 MB/s eta 0:00:00
#12 4.319 Collecting filelock<4,>=3.12.2
#12 4.327   Downloading filelock-3.17.0-py3-none-any.whl (16 kB)
#12 4.349 Collecting distlib<1,>=0.3.7
#12 4.356   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.368      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 48.1 MB/s eta 0:00:00
#12 4.404 Collecting platformdirs<5,>=3.9.1
#12 4.412   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#12 4.495 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 4.677 Successfully installed distlib-0.3.9 filelock-3.17.0 platformdirs-4.3.6 virtualenv-20.29.2
#12 DONE 4.8s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      128960 kB
DirectMap2M:     7211008 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2025-02-18/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
##[endgroup]
fmt check
Diff in /checkout/compiler/rustc_pattern_analysis/src/tests.rs:456:
 
 /// Analyze a match made of these patterns.
 fn check_exhaustiveness(patterns: Vec<DeconstructedPat<Cx>>) -> Vec<WitnessPat<Cx>> {
-    let report = check(&patterns, usize::MAX) .unwrap();
+    let report = check(&patterns, usize::MAX).unwrap();
     report.non_exhaustiveness_witnesses
 
Diff in /checkout/compiler/rustc_pattern_analysis/src/tests.rs:537:
 
 /// Analyze a match made of these patterns and returns the computed arm intersections.
 /// Analyze a match made of these patterns and returns the computed arm intersections.
 fn check_intersections(patterns: Vec<DeconstructedPat<Cx>>) -> Vec<Vec<usize>> {
-    let report = check(&patterns, usize::MAX) .unwrap();
+    let report = check(&patterns, usize::MAX).unwrap();
     report.arm_intersections.into_iter().map(|bitset| bitset.iter().collect()).collect()
 
fmt: checked 5863 files
Build completed unsuccessfully in 0:01:12
  local time: Mon Mar  3 11:43:29 UTC 2025
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13638635282/job/38123286503) [(plain)](https://github.com/rust-lang/rust/commit/fe01d30fb3a68b13c65b3e0f22ecefa5b768498c/checks/38123286503/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #138267) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #138310) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote commented:
https://github.com/rust-lang/compiler-team/issues/844 has hit 10 days after seconding without objection, so is approved. And we have established `RUSTC_LINT_FLAGS` as the best way to handle new lints for compiler crates. So this is ready to go.

@bors rollup

nnethercote commented on compiler/rustc_pattern_analysis/src/lib.rs:
```diff
@@ -19,6 +19,9 @@ pub mod pat_column;
 pub mod rustc;
 pub mod usefulness;
 
+#[cfg(test)]
+mod tests;
+
```
I don't like having a "dual status" crate like this. I think it should either be a compiler crate and treated like other compiler crates, or be on crates.io (like the `ena` crate, for example) and then it can do things differently.

Nonetheless, the third commit it not integral and can be removed. Some level of false positive workaround is necessary for the first commit, such as the `#![allow(unused_crate_dependencies)]` attributes.

Nadrieril replied:
It still evolves sufficiently in sync with the compiler that moving out-of-tree wouldn't make sense. I expect other librarified parts of the compiler to behave similarly, how do they behave? The only other one I know of is rustc_type_ir.

nnethercote replied:
`rustc_transmute` also has a `rustc` feature.

`rustc_abi`, `rustc_next_trait_solver`, `rustc_index`, `rustc_ast_ir`, `rustc_transmute`, `rustc_type_ir` all have a `nightly` feature.

I don't have a sense of whether `rustc` and `nightly` are doing the same thing. I wonder if some of these crates could be split into the "base" part (typically shared with rust-analyzer) and the rustc-specific part.

Nadrieril replied:
`rustc_pattern_analysis` could easily be split in that way. Looking at these crates, I see:
- actual nightly features (e.g. `Rc::new_zeroed` or `Extend::extend_one`);
- implementing `Encodable`/`Decodable`;
- whole submodules cfged-out;
- internal features like `rustc_pass_by_value` or `rustc_lint_query_instability` (only in `rustc_type_ir`).

My guess is that we could split most of them if we had the orphan rules escape hatch that RfL wants.

jieyouxu commented:
Commit 1 LGTM. I'll hand the review over to @Nadrieril who has a stronger preference on the test organizations.

r? @Nadrieril

bors commented:
:umbrella: The latest upstream changes (presumably #138464) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote commented:
I rebased and removed the third commit changing the `rustc_pattern_analysis` tests.

bors commented:
:umbrella: The latest upstream changes (presumably #137011) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Nadrieril commented:
r=me after rebase

nnethercote commented:
I rebased. Amusingly, in the time this PR has been alive several new unnecessary dependencies have been introduced. I think I've had to remove one or two every time I rebased.

@bors r=jieyouxu,Nadrieril

bors commented:
:pushpin: Commit 1fa77ca45e6771f51e987cc5ca55fa52ecdb4988 has been approved by `jieyouxu,Nadrieril`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu,Nadrieril 1fa77ca45e6771f51e987cc5ca55fa52ecdb4988 -->
<!-- homu: {"type":"Approved","sha":"1fa77ca45e6771f51e987cc5ca55fa52ecdb4988","approver":"jieyouxu,Nadrieril","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:evergreen_tree: The tree is currently [closed](https://github.com/rust-lang/rust/pull/138653#issuecomment-2733977420) for pull requests below priority 100. This pull request will be tested once the tree is reopened.

bors commented:
:umbrella: The latest upstream changes (presumably #138693) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote commented:
I rebased again, and again removed a couple of new unnecessary dependencies that had crept in.

@bors r=jieyouxu,Nadrieril p=1

Bumping priority because this is conflict-prone.

bors commented:
:pushpin: Commit f84fd39dd0e08037493489ee1c6f10b8d2f44978 has been approved by `jieyouxu,Nadrieril`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu,Nadrieril f84fd39dd0e08037493489ee1c6f10b8d2f44978 -->
<!-- homu: {"type":"Approved","sha":"f84fd39dd0e08037493489ee1c6f10b8d2f44978","approver":"jieyouxu,Nadrieril","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

nnethercote commented:
@bors rollup=never

bors commented:
:hourglass: Testing commit f84fd39dd0e08037493489ee1c6f10b8d2f44978 with merge cea7074d252b8de75b39c425e8305c1b603c38bf...
<!-- homu: {"type":"BuildStarted","head_sha":"f84fd39dd0e08037493489ee1c6f10b8d2f44978","merge_sha":"cea7074d252b8de75b39c425e8305c1b603c38bf"} -->

rust-log-analyzer commented:
The job **`dist-x86_64-musl`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/13956484361/job/39068813155) [(plain)](https://github.com/rust-lang-ci/rust/commit/cea7074d252b8de75b39c425e8305c1b603c38bf/checks/39068813155/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13956484361/job/39071380431)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/13956484361/job/39071380431","builder_name":"checks-actions"} -->

nnethercote commented:
@bors r=jieyouxu,Nadrieril p=1

bors commented:
:pushpin: Commit e2320b32c5268913349549fc59972443975ceec4 has been approved by `jieyouxu,Nadrieril`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu,Nadrieril e2320b32c5268913349549fc59972443975ceec4 -->
<!-- homu: {"type":"Approved","sha":"e2320b32c5268913349549fc59972443975ceec4","approver":"jieyouxu,Nadrieril","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

jieyouxu commented:
@bors p=10 (or else this will get rollup-raced)

bors commented:
:hourglass: Testing commit e2320b32c5268913349549fc59972443975ceec4 with merge 4e2b096ed6c8a1400624a54f6c4fd0c0ce48a579...
<!-- homu: {"type":"BuildStarted","head_sha":"e2320b32c5268913349549fc59972443975ceec4","merge_sha":"4e2b096ed6c8a1400624a54f6c4fd0c0ce48a579"} -->

rust-timer commented:
Finished benchmarking commit (4e2b096ed6c8a1400624a54f6c4fd0c0ce48a579): [comparison URL](https://perf.rust-lang.org/compare.html?start=70237a8cb905fc48ed6148d2f2dba756535a5940&end=4e2b096ed6c8a1400624a54f6c4fd0c0ce48a579&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=70237a8cb905fc48ed6148d2f2dba756535a5940&end=4e2b096ed6c8a1400624a54f6c4fd0c0ce48a579&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=70237a8cb905fc48ed6148d2f2dba756535a5940&end=4e2b096ed6c8a1400624a54f6c4fd0c0ce48a579&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=70237a8cb905fc48ed6148d2f2dba756535a5940&end=4e2b096ed6c8a1400624a54f6c4fd0c0ce48a579&stat=cycles:u)
<details>
<summary>Results (primary -3.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -3.0% | [-3.0%, -3.0%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -3.0% | [-3.0%, -3.0%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=70237a8cb905fc48ed6148d2f2dba756535a5940&end=4e2b096ed6c8a1400624a54f6c4fd0c0ce48a579&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 773.52s -> 773.468s (-0.01%)
**Artifact size**: 365.50 MiB -> 365.48 MiB (-0.01%)
